crates/qasm3/src/circuit.rs

```diff
@@ -86,30 +83,38 @@ pub struct PyGate {
 }
 
 impl PyGate {
-    pub fn new<T: IntoPy<Py<PyAny>>, S: AsRef<str>>(
-        py: Python,
+    pub fn new<
+        'py,
+        T: IntoPyObject<'py, Target = PyAny, Output = Bound<'py, PyAny>>,
+        S: AsRef<str>,
+    >(
+        py: Python<'py>,
         constructor: T,
         name: S,
         num_params: usize,
         num_qubits: usize,
-    ) -> Self {
+    ) -> Self
+    where
+        <T as pyo3::IntoPyObject<'py>>::Error: std::fmt::Debug,
+    {
         Self {
-            constructor: constructor.into_py(py),
+            constructor: constructor.into_pyobject(py).unwrap().unbind(),
             name: name.as_ref().to_owned(),
             num_params,
             num_qubits,
         }
     }
 
     /// Construct a Python-space instance of the custom gate.
-    pub fn construct<A>(&self, py: Python, args: A) -> PyResult<Py<PyAny>>
+    pub fn construct<'py, A>(&'py self, py: Python<'py>, args: A) -> PyResult<Py<PyAny>>
     where
-        A: IntoPy<Py<PyTuple>>,
+        A: pyo3::IntoPyObject<'py, Target = PyTuple, Output = Bound<'py, PyTuple>>,
+        <A as pyo3::IntoPyObject<'py>>::Error: std::fmt::Debug,
     {
-        let args = args.into_py(py);
-        let received_num_params = args.bind(py).len();
+        let args = args.into_pyobject(py).unwrap();
```

Comment:
This function already returns `PyResult`, so no need for the crazy trait bounds and `unwrap`, I think?

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/13577/commits/0577f48a7814455551ec2cda3540a9d1a7ad9847